crates/runtime/src/instance/allocator/pooling.rs

```diff
@@ -0,0 +1,1714 @@
+    /// runtime bounds checks.
+    ///
+    /// On 32-bit platforms, the default for this value will be 10 MiB. A 32-bit host has very limited address
+    /// space to reserve for a lot of concurrent instances.  As a result, runtime bounds checking will be used
+    /// for all memory accesses.  For better runtime performance, a 64-bit host is recommended.
+    ///
+    /// This value will be rounded up by the WebAssembly page size (64 KiB).
+    pub memory_reservation_size: u64,
+}
+
+impl Default for InstanceLimits {
+    fn default() -> Self {
+        // See doc comments for `InstanceLimits` for these default values
+        Self {
+            count: 1000,
+            #[cfg(target_pointer_width = "32")]
+            memory_reservation_size: 0xA00000,
+            #[cfg(target_pointer_width = "64")]
+            memory_reservation_size: 0x180000000,
+        }
+    }
+}
+
+/// The allocation strategy to use for the pooling instance allocator.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum PoolingAllocationStrategy {
+    /// Allocate from the next available instance.
+    NextAvailable,
+    /// Allocate from a random available instance.
+    Random,
+}
+
+impl PoolingAllocationStrategy {
+    fn next(&self, free_count: usize) -> usize {
+        debug_assert!(free_count > 0);
+
+        match self {
+            Self::NextAvailable => free_count - 1,
+            Self::Random => rand::thread_rng().gen_range(0, free_count),
+        }
+    }
+}
+
+impl Default for PoolingAllocationStrategy {
+    fn default() -> Self {
+        Self::NextAvailable
+    }
+}
+
+// Used to iterate the base address of instance memories and tables.
+struct BasePointerIterator {
+    base: *mut u8,
+    current: usize,
+    num: usize,
+    size: usize,
+}
+
+impl BasePointerIterator {
+    fn new(base: *mut u8, num: usize, size: usize) -> Self {
+        Self {
+            base,
+            current: 0,
+            num,
+            size,
+        }
+    }
+}
+
+impl Iterator for BasePointerIterator {
+    type Item = *mut u8;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let current = self.current;
+        if current == self.num {
+            return None;
+        }
+
+        self.current += 1;
+
+        Some(unsafe { self.base.add(current * self.size) })
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let remaining = self.num - self.current;
+        (remaining, Some(remaining))
+    }
+}
+
+/// Represents a pool of maximal `Instance` structures.
+///
+/// Each index in the pool provides enough space for a maximal `Instance`
+/// structure depending on the limits used to create the pool.
+///
+/// The pool maintains a free list for fast instance allocation.
+///
+/// The userfault handler relies on how instances are stored in the mapping,
+/// so make sure the uffd implementation is kept up-to-date.
+#[derive(Debug)]
+struct InstancePool {
+    mapping: Mmap,
```

Comment:
I think it would make sense to make this more generic especially if more free lists are maintained per my last comment.

The mutex would probably then surround the instance pool itself rather than its free list, as we only want one lock to be able to modify three different free lists (instance, memory, and table).

Reply:
And yeah I wouldn't really go out of your way to try to make this generic and wrapped up nicely, only if it makes sense. Just something I noticed while reading!

Reply:
While `MemoryPool` and `TablePool` have a lot of overlap, they differ a bit from `InstancePool` and `StackPool` to be entirely generic.

Consolidating `MemoryPool` and `TablePool` probably isn't worth it as the impls are pretty small.

Resolving this as "won't fix" since we're not changing `MemoryPool` and `TablePool` to maintain their own free lists.